pkg/kafka/request.go

```diff
@@ -0,0 +1,126 @@
+// Copyright 2017 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package kafka
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+
+	"github.com/optiopay/kafka/proto"
+)
+
+// RequestMessage represents a Kafka request message
+type RequestMessage struct {
+	kind    int16
+	version int16
+	rawMsg  []byte
+	request interface{}
+}
+
+// GetKind returns the kind of Kafka request
+func (req *RequestMessage) GetKind() int16 {
+	return req.kind
+}
+
+// GetRaw returns the raw Kafka request
+func (req *RequestMessage) GetRaw() []byte {
+	return req.rawMsg
+}
+
+// String returns a human readable representation of the request message
+func (req *RequestMessage) String() string {
+	b, err := json.Marshal(req.request)
+	if err != nil {
+		return err.Error()
+	}
+	return fmt.Sprintf("apiKey=%d,apiVersion=%d,len=%d: %s",
+		req.kind, req.version, len(req.rawMsg), string(b))
+}
+
+// CreateResponse creates a response message based on the provided request
+// message. The response will have the specified error code set in all topics
+// and embedded partitions.
+func (req *RequestMessage) CreateResponse(err error) (*ResponseMessage, error) {
+	switch val := req.request.(type) {
+	case *proto.ProduceReq:
+		return createProduceResponse(val, err)
+	case *proto.FetchReq:
+		return createFetchResponse(val, err)
+	case *proto.OffsetReq:
+		return createOffsetResponse(val, err)
+	case *proto.MetadataReq:
+		return createMetadataResponse(val, err)
+	case *proto.ConsumerMetadataReq:
+		return createConsumerMetadataResponse(val, err)
+	case *proto.OffsetCommitReq:
+		return createOffsetCommitResponse(val, err)
+	case *proto.OffsetFetchReq:
+		return createOffsetFetchResponse(val, err)
+	}
+
+	return nil, fmt.Errorf("unknown request")
```

Comment:
`s/fmt.Errorf("unknown request")/fmt.Errorf("unknown request: %s", err)/`?

Reply:
I think it should log `req.request` or `req.rawMsg` or `req.kind`.

Reply:
That doesn't make any sense. It's an unknown type, err has a different context here.

Reply:
Added numeric value to error message